eZ/Publish/Core/Persistence/Legacy/Content/Mapper.php

```diff
@@ -176,64 +181,146 @@ public function convertToStorageValue(Field $field)
             }
 
-            $fieldId = (int)$row['ezcontentobject_attribute_id'];
-            if (!isset($fields[$contentId][$versionId][$fieldId])) {
+            $fieldId = (int)$row["{$prefix}attribute_id"];
+            $fieldDefinitionId = (int)$row["{$prefix}attribute_contentclassattribute_id"];
+            $languageCode = $row["{$prefix}attribute_language_code"];
+
+            if (!isset($fields[$contentId][$versionId][$fieldId])
+                && isset($fieldDefinitions[$contentId][$versionId][$languageCode][$fieldDefinitionId])
+            ) {
                 $fields[$contentId][$versionId][$fieldId] = $this->extractFieldFromRow($row);
+                unset($fieldDefinitions[$contentId][$versionId][$languageCode][$fieldDefinitionId]);
             }
         }
 
+        return $this->buildContentObjects(
+            $contentInfos,
+            $versionInfos,
+            $fields,
+            $fieldDefinitions,
+            $versionedNameData
+        );
+    }
+
+    /**
+     * @return \eZ\Publish\SPI\Persistence\Content[]
+     */
+    private function buildContentObjects(
+        array $contentInfos,
+        array $versionInfos,
+        array $fields,
+        array $fieldDefinitions,
+        array $versionedNameData
+    ): array {
         $results = [];
+
         foreach ($contentInfos as $contentId => $contentInfo) {
             foreach ($versionInfos[$contentId] as $versionId => $versionInfo) {
                 // Fallback to just main language name if versioned name data is missing
-                if (isset($versionedNameData[$contentId][$versionInfo->versionNo])) {
-                    $names = $versionedNameData[$contentId][$versionInfo->versionNo];
-                } else {
-                    $names = [$contentInfo->mainLanguageCode => $contentInfo->name];
-                }
+                $names = $versionedNameData[$contentId][$versionInfo->versionNo]
+                    ?? [$contentInfo->mainLanguageCode => $contentInfo->name];
 
                 $content = new Content();
                 $content->versionInfo = $versionInfo;
                 $content->versionInfo->names = $names;
                 $content->versionInfo->contentInfo = $contentInfo;
                 $content->fields = array_values($fields[$contentId][$versionId]);
+
+                foreach ($fieldDefinitions[$contentId][$versionId] as $languageCode => $versionFieldDefinitions) {
+                    foreach ($versionFieldDefinitions as $fieldDefinition) {
+                        $content->fields[] = $this->createEmptyField(
+                            $fieldDefinition,
+                            $languageCode
+                        );
+                    }
+                }
+
                 $results[] = $content;
             }
         }
 
         return $results;
     }
 
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
```

Comment:
missing `@return` type array shape and `@param array $rows` argument array-shape. 
We don't have PHPStan in kernel 1.3 to report that :(